builds.go

```diff
@@ -128,9 +198,97 @@ func resolveRemoteArch(ctx context.Context, img string, arch string) (*RemoteRes
 	return resolved, nil
 }
 
+type cacheFileContents struct {
+	Resolve map[string]*cacheResolveType `json:"resolve"`
+}
+
+var (
+	saveCache      *cacheFileContents
+	saveCacheMutex sync.Mutex
+)
+
+func loadCacheFromFile() error {
+	if cacheFile == "" {
+		return nil
+	}
+
+	// now that we know we have a file we want cache to go into (and come from), let's initialize the "saveCache" (which will be written when the whole process is done / we're successful, and *only* caches staging images)
+	saveCacheMutex.Lock()
+	saveCache = &cacheFileContents{Resolve: map[string]*cacheResolveType{}}
+	saveCacheMutex.Unlock()
+
+	f, err := os.Open(cacheFile)
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	var cache cacheFileContents
+	err = json.NewDecoder(f).Decode(&cache) // *technically*, this will silently ignore garbage (or extra documents) at the end of the file, but it's for our cache file so it's not really an issue for us (the only input to this should be our own output)
+	if err != nil {
+		return err
+	}
+
+	for img, r := range cache.Resolve {
+		r := r // https://github.com/golang/go/issues/60078
+		fun, _ := cacheResolve.LoadOrStore(img, sync.OnceValues(func() (*cacheResolveType, error) {
+			return r, nil
+		}))
+		r2, err := fun.(func() (*cacheResolveType, error))()
+		if err != nil {
+			// this should never happen (hence panic vs return) 🙈
+			panic(err)
+		}
+		if r2 != r {
+			panic("r2 != r??? " + img)
+		}
+	}
+
+	return nil
+}
+
+func saveCacheToFile() error {
+	saveCacheMutex.Lock()
+	defer saveCacheMutex.Unlock()
+
+	if saveCache == nil || cacheFile == "" {
```

Comment:
Could this be moved before locking the mutex?

Reply:
The `cacheFile` test could, but the test for `saveCache` cannot because we cannot safely compare that value without a mutex (even a simple comparison like `nil`).

(This is also effectively the test for "did the user specify `--cache`", and we're specifying `--cache` in all the places we care about now, so this test is effectively as good as a no-op now.)